core/src/main/java/io/questdb/std/LongList.java

```diff
@@ -100,74 +100,136 @@ public void add(LongList that, int lo, int hi) {
         int p = pos;
         int s = hi - lo;
         ensureCapacity(p + s);
-        System.arraycopy(that.buffer, lo, this.buffer, p, s);
+        System.arraycopy(that.data, lo, this.data, p, s);
         pos += s;
     }
 
     public void add(int index, long element) {
         ensureCapacity(++pos);
-        System.arraycopy(buffer, index, buffer, index + 1, pos - index - 1);
-        buffer[index] = element;
+        System.arraycopy(data, index, data, index + 1, pos - index - 1);
+        data[index] = element;
     }
 
     public void arrayCopy(int srcPos, int dstPos, int length) {
-        System.arraycopy(buffer, srcPos, buffer, dstPos, length);
+        System.arraycopy(data, srcPos, data, dstPos, length);
     }
 
-    public int binarySearch(long v) {
+    public int binarySearch(long value, int scanDir) {
         int low = 0;
         int high = pos;
 
+        if (low == high) {
+            return -1;
+        }
+
         while (low < high) {
 
             if (high - low < 65) {
-                return scanSearch(v, low, high);
+                return scanSearch(value, low, high);
             }
 
             int mid = (low + high - 1) >>> 1;
-            long midVal = buffer[mid];
-
-            if (midVal < v)
-                low = mid + 1;
-            else if (midVal > v)
+            long midVal = data[mid];
+
+            if (midVal < value) {
+                if (low < mid) {
+                    low = mid;
+                } else {
+                    if (data[high] > value) {
+                        return -low - 1;
+                    }
+                    return -high - 1;
+                }
+            } else if (midVal > value)
                 high = mid;
-            else
-                return mid;
+            else {
+                // In case of multiple equal values, find the first
+                mid += scanDir;
```

Comment:
this is a direct port of throughly tested (by O3) implementation found in `util.h`. It is quite time consuming to test and I feel like we're having to test copies of code Java is not able to reuse. In ideal world I would not have a copy of binary search here, but since it is here, I would rather add a reference on the original implementation and a note that they have to be in sync, always.